vscode: add possibility to add command Line arguments

heijligen opened the pull request:
This can be used to run vscode under wayland when it is published
with electron 12. The arguments will be passt to electron and
further to chromium.
Based on https://nixos.org/manual/nixpkgs/stable/#ssec-gnome-hooks

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Be able to use vscode nativly under wayland. Use the same behavior as chromium

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

turion commented:
Sounds plausible, but I don't have a wayland to test it (and not planning on installing it right now). Can you parametrize and extend `nixos/tests/vscodium.nix` such that there are two tests, one with X and the other with wayland? And then test that the wayland vscodium test doesn't work in the VM without your change, and it does work with your change?

turion commented:
Heads up, possible (but improbable) conflict: https://github.com/NixOS/nixpkgs/pull/118292#issuecomment-812264089

r-rmcgibbo commented:
Result of `nixpkgs-review pr 118360` at 94fbe7cc run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built successfully:</summary>
  <ul>
    <li>vscode</li>
    <li>vscode-with-extensions</li>
    <li>vscodium</li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

wrapGAppsHook is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/applications/editors/vscode/generic.nix:64:5:
```
   |
64 |     buildInputs = [ libsecret libXScrnSaver libxshmfence ]
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/applications/editors/vscode/generic.nix:74:5:
```
   |
74 |     installPhase =
   |     ^
```
</li>
  </ul>
</details>

heijligen commented:
Currently no released vscode / vscodium features wayland support. It will need electron 12. I've my own vscode (code-oss) build with electron 12 patches and wayland support. https://github.com/microsoft/vscode/pull/120034
The option is for passing any command line argument to vscode -> electron -> chromium. Wayland support is only my usecase for this.

heijligen commented:
There is no environment for wayland tests in nixos/tests. Chromium let the user also just add commandLineArgs

jul1u5 commented:
VS Code 1.56 is now shipped with Electron 12 and running `code --enable-features=UseOzonePlatform --ozone-platform=wayland` uses Wayland. It would be great to have this merged.

turion commented:
> There is no environment for wayland tests in nixos/tests

What do you mean by that? I thought you can just enter arbitrary config for the virtual machines, so surely you can configure the test VM to run Wayland?

If you mean something like this:
```nix
    imports = [
      ./common/x11.nix
    ];
```
You could add a file `./common/wayland.nix` with a minimalistic wayland config.

heijligen commented:
Wayland is only a protocol. There is no central server component like the X Server. Each Compositor like Gnomes Mutter, Kdes Kwin or sway implement the server side of the wayland protocol. Tests would be needed for each one.

Beside that, this PR adds the possibility to pass any command line arguments to VS Code -> Electron -> Chromium and is not Wayland specific. So I see no need for testing an specific overrode VS Code. 

My personal intension behind this PR is to use the generic usable commandLineArgs attribute to pass the same arguments to the Chromium under VS Code I'm passing to my normal Chromium

/etc/nixos/configuration.nix: ```
[...]
chromium = (pkgs.chromium.override {
      commandLineArgs =
        "--enable-features=VaapiVideoDecoder,UseOzonePlatform " +
        "--enable-accelerated-video-decode " +
        "--ozone-platform=wayland " +
        "--enable-gpu-rasterization " +
        "--disable-sync-preferences " +
        "--force-dark-mode " +
        "--enable-native-notifications";
    });

vscode = (pkgs.vscode.override {
      commandLineArgs =
        "--enable-features=VaapiVideoDecoder,UseOzonePlatform " +
        "--enable-accelerated-video-decode " +
        "--ozone-platform=wayland " +
        "--enable-gpu-rasterization " +
        "--force-dark-mode " +
        "--enable-native-notifications";
    });
```

turion commented:
As you like. I'm just recommending a way for you to ensure that noone in the future will break the feature you're introducing.

kisik21 commented:
For testing, you just need to ensure that a wayland compositor (any wayland compositor, I suggest sway as it's easy to install and launch, just one option) is running and VSCode (or whatever you want to test) successfully launches with `WAYLAND_DISPLAY` set and `DISPLAY` unset.

Additionally, you CAN launch Sway inside of an X server - you'll need to manage some variables manually but it is possible, as Sway can be launched with an X11 backend instead of a DRM one. It'll look like a window on your desktop with a second desktop, kinda like a remote desktop, but locally.

heijligen commented:
I can have a look on a sway testing environment. But would do this outside from this PR

Synthetica9 commented:
I don't think we should pollute the sway test with all sorts of random programs... Maybe we could make a template to test programs in cage instead?

Synthetica9 commented:
Now done here: https://github.com/NixOS/nixpkgs/pull/123680

kisik21 commented:
What's blocking merging this and #123680? I would love to see this in `nixos-unstable`.

SuperSandro2000 commented on pkgs/applications/editors/vscode/generic.nix:
```diff
@@ -93,5 +94,11 @@ in
       '';
 
+    preFixup = ''
+      eval gappsWrapperArgs+=(
```
```suggestion
      gappsWrapperArgs+=(
```

mohe2015 commented:
@heijligen can you rebase this please? (and apply the suggestion if it works)

heijligen committed:
vscode: add possibility to add command Line arguments

This can be used to run vscode under wayland when it is published
with electron 12. The arguments will be passt to electron and
further to chromium.
Based on https://nixos.org/manual/nixpkgs/stable/#ssec-gnome-hooks

Co-authored-by: Moritz Hedtke <Moritz.Hedtke@t-online.de>
